### DISCUSSION TITLE
How to use winget to install VSCode with custom options?

### DISCUSSION THREAD

**Discussion by Zamiell** [microsoft/winget-cli#1798] [Q&A] [answered]
Hello friends. Thanks so much for winget. I have a quick question.

I know that I can install Visual Studio Code with winget by typing: `winget install vscode`

However, doing this will install Visual Studio Code with all default options, which is not what I want.

In the Visual Studio Code GUI installer, there is a screen that has a checkbox for adding Visual Studio Code to the right-click context menu, which is a very important feature:

![Image of VSCode GUI installer](https://user-images.githubusercontent.com/9638156/48968771-6f715080-f005-11e8-8486-220cc183a830.png)

These checkboxes are not checked by default. And `winget` does not seem to check them when it executes the installer.

Is there a way to pass custom installer arguments to winget? And if so, how would I even go about finding out what the correct installer arguments to pass even are?

## January 2026 Edit

This question is quite old now. The below answers detail how to install it with custom options. However, the complexity of this is high. I think a "cleaner" solution is to just install Visual Studio Code the "normal" way with `winget`, and then add the context menu afterward with the following .reg file:

```reg
Windows Registry Editor Version 5.00

; By default, when VSCode is installed from winget, it does not include context menus, so it must be
; manually added afterwards.
; https://thisdavej.com/right-click-on-windows-folder-and-open-with-visual-studio-code/

; Open files
[HKEY_CLASSES_ROOT\*\shell\Open with VS Code]
@="Open with Code"
"Icon"="C:\\Users\\james\\AppData\\Local\\Programs\\Microsoft VS Code\\Code.exe,0"
[HKEY_CLASSES_ROOT\*\shell\Open with VS Code\command]
@="\"C:\\Users\\james\\AppData\\Local\\Programs\\Microsoft VS Code\\Code.exe\" \"%1\""

; This will make it appear when you right click ON a folder
; The "Icon" line can be removed if you don't want the icon to appear
[HKEY_CLASSES_ROOT\Directory\shell\vscode]
@="Open with Code"
"Icon"="\"C:\\Users\\james\\AppData\\Local\\Programs\\Microsoft VS Code\\Code.exe\",0"
[HKEY_CLASSES_ROOT\Directory\shell\vscode\command]
@="\"C:\\Users\\james\\AppData\\Local\\Programs\\Microsoft VS Code\\Code.exe\" \"%1\""

; This will make it appear when you right click INSIDE a folder
; The "Icon" line can be removed if you don't want the icon to appear
[HKEY_CLASSES_ROOT\Directory\Background\shell\vscode]
@="Open with Code"
"Icon"="\"C:\\Users\\james\\AppData\\Local\\Programs\\Microsoft VS Code\\Code.exe\",0"
[HKEY_CLASSES_ROOT\Directory\Background\shell\vscode\command]
@="\"C:\\Users\\james\\AppData\\Local\\Programs\\Microsoft VS Code\\Code.exe\" \"%V\""
```

**Comment by jedieaston** [comment, marked as answer]
You can use the `--override` argument on `winget install` to replace the arguments winget uses with ones you want yourself. Unfortunately, you'll probably have to do some googling or run the installer with `/?` to get all possible arguments. 

To install VSCode with those options selected, run `winget install Microsoft.VisualStudioCode --override '/SILENT /mergetasks="!runcode,addcontextmenufiles,addcontextmenufolders"'`

**Comment by haltbarmilch** [comment]
Hello! Will the `--override` parameter be saved for an upgrade? Or must i pass the argument on every upgrade? Can't find any info on this. Thanks a lot.

**Reply by jedieaston** [reply]
It will not be saved, unfortunately, but on upgrade VSCode knows to keep the shortcuts in PATH. It should still work the way you'd expect.

**Reply by sanamhub** [reply]
The answer is No, Me neither didn't find the exact answer on the Internet btw, I come say the answer is No because I did tested on my System :)

**Reply by dominik-ba** [reply]
I'd be tempted to say these parameters are only happening once during the installation and therefore is nothing that needs to be done again and again during an upgrade.

**Comment by kellyelton** [comment]
What about those other options, what are the tasks for those?

**Reply by kellyelton** [reply]
https://github.com/microsoft/vscode/blob/main/build/win32/code.iss#L76

```
[Tasks]
Name: "desktopicon"; Description: "{cm:CreateDesktopIcon}"; GroupDescription: "{cm:AdditionalIcons}"; Flags: unchecked
Name: "quicklaunchicon"; Description: "{cm:CreateQuickLaunchIcon}"; GroupDescription: "{cm:AdditionalIcons}"; Flags: unchecked; OnlyBelowVersion: 0,6.1
Name: "addcontextmenufiles"; Description: "{cm:AddContextMenuFiles,{#NameShort}}"; GroupDescription: "{cm:Other}"; Flags: unchecked
Name: "addcontextmenufolders"; Description: "{cm:AddContextMenuFolders,{#NameShort}}"; GroupDescription: "{cm:Other}"; Flags: unchecked; Check: not (IsWindows11OrLater and QualityIsInsiders)
Name: "associatewithfiles"; Description: "{cm:AssociateWithFiles,{#NameShort}}"; GroupDescription: "{cm:Other}"
Name: "addtopath"; Description: "{cm:AddToPath}"; GroupDescription: "{cm:Other}"
Name: "runcode"; Description: "{cm:RunAfter,{#NameShort}}"; GroupDescription: "{cm:Other}"; Check: WizardSilent
```

**Comment by Larandar** [comment]
So after some tinkering of everything in this thread, and some trial and error (might have needed to restart my computer 3 time and uninstall vscode 10 times) I found a "good" set of options that make vscode install silent in winget:

```pwsh
.\VSCodeUserSetup-x64-1.85.0.exe /VERYSILENT /SP- /MERGETAKS="!runcode,!desktopicon,addcontextmenufiles,addcontextmenufolders,associatewithfiles,addtopath"
```

- `/VERYSILENT /SP-` make it so no window is opened (unless you have a code instance opened already)
- The `/CLOSEAPPLICATIONS` and `FORCECLOSEAPPLICATIONS` do not work from what I tried, so you still get a popu asking you to quit vs-code
- `/MERGETASKS=!runcode` prevent vs-code to open after the install
- `/MERGETASKS=!desktopicon` do not add desktop icon (already present in the Win menu)
- `/MERGETASKS=addcontextmenufiles,addcontextmenufolders,associatewithfiles,addtopath` does all the background stuffs that are missing (weirdly Ninite use WinGet to install VS-Code addtopath but not the context menu for folder?)

This all make it into my dotfile bootstrap as:
```pwsh
Write-Host -ForegroundColor Yellow "🌋 Force reinstall of VS-Code to ensure Path and Shell integration"
winget install --force Microsoft.VisualStudioCode --override '/VERYSILENT /SP- /MERGETAKS="!runcode,!desktopicon,addcontextmenufiles,addcontextmenufolders,associatewithfiles,addtopath"'
```

**Reply by andregeuze** [reply]
MERGETAKS looks like a typo, why DOES it work though?

**Reply by neurocline** [reply]
I know many command-line parsers allow unique prefixes to work; maybe there's a bug where it sees the unique prefix and just stops parsing (e.g. it sees "/MER" and says "aha, that's /MERGETASKS".

**Comment by emilwojcik93** [comment]
# Fixing Missing Context Menu Entries in VS Code Installed via Winget

## TL;DR - Working Command

```powershell
winget install --force Microsoft.VisualStudioCode --override '/VERYSILENT /SP- /MERGETASKS="addcontextmenufiles,addcontextmenufolders,associatewithfiles,addtopath"'
```

## The Issue & Solution

Many users experience missing "Open with Code" options in context menus after installing VS Code via winget. This comprehensive guide explains the issue and provides solutions for proper installation.

### Root Cause Identified

The problem occurs due to a parameter typo in a commonly shared command:

```powershell
winget install --force Microsoft.VisualStudioCode --override '/VERYSILENT /SP- /MERGETAKS="!runcode,!desktopicon,addcontextmenufiles,addcontextmenufolders,associatewithfiles,addtopath"'
```

Note the misspelling: `MERGETAKS` instead of `MERGETASKS` (missing the "S"). 

This subtle typo has significant consequences: the VS Code installer silently ignores unrecognized parameters without displaying any error, resulting in default installation settings that skip context menu integration.

## Parameter Explanation

Let's break down the correct command:

### Winget Parameters

| Parameter | Purpose |
|:----------|:--------|
| `install` | Basic command to install a package |
| `Microsoft.VisualStudioCode` | Package identifier for VS Code |
| `--override` | Passes arguments directly to the installer |

### VS Code Installer Parameters

| Parameter | Purpose |
|:----------|:--------|
| `/VERYSILENT` | Runs installation without UI |
| `/SP-` | Suppresses initial setup message |
| `/MERGETASKS` | **Critical parameter** that specifies installation tasks |

### Key MERGETASKS Options

| Option | Function | Impact |
|:-------|:---------|:-------|
| `addcontextmenufiles` | Adds context menu entry for files | Enables right-click "Open with Code" on files |
| `addcontextmenufolders` | Adds context menu entry for folders | Enables right-click "Open with Code" on folders |
| `associatewithfiles` | Associates VS Code with supported file types | Makes VS Code the default editor for these files |
| `addtopath` | Adds VS Code to PATH | Enables using `code` command in terminal |

## How to Fix an Existing Installation

If you've already installed VS Code without context menu integration:

1. Uninstall VS Code:
(optional step `winget install --force` should reinstall target package)
   ```powershell
   winget uninstall Microsoft.VisualStudioCode
   ```

2. Reinstall with the correct parameters:
   ```powershell
   winget install --force Microsoft.VisualStudioCode --override '/VERYSILENT /SP- /MERGETASKS="addcontextmenufiles,addcontextmenufolders,associatewithfiles,addtopath"'
   ```

## Installation Options for Different Needs

### Standard Installation (Recommended)
```powershell
winget install --force Microsoft.VisualStudioCode --override '/VERYSILENT /SP- /MERGETASKS="addcontextmenufiles,addcontextmenufolders,associatewithfiles,addtopath"'
```

### Complete Installation with All Features
```powershell
winget install --force Microsoft.VisualStudioCode --override '/VERYSILENT /SP- /MERGETASKS="runcode,desktopicon,addcontextmenufiles,addcontextmenufolders,associatewithfiles,addtopath"'
```

### Minimal Installation with Context Menu Only
```powershell
winget install --force Microsoft.VisualStudioCode --override '/VERYSILENT /SP- /MERGETASKS="!runcode,!desktopicon,addcontextmenufiles,addcontextmenufolders,!associatewithfiles,addtopath"'
```

## Advanced: Understanding MERGETASKS Behavior

- Tasks without a prefix are explicitly selected
- Tasks prefixed with `!` are explicitly deselected
- Default VS Code installer behavior is to select most options

Example: `/MERGETASKS="!runcode,addcontextmenufiles"` will disable automatic launch of VS Code after installation but ensure file context menu integration.

## Available VS Code Installer Options

<details>
<summary>Click to view all available installer options</summary>

<img src="https://github.com/user-attachments/assets/85ab97e2-2ac3-44ce-bf5e-3739c1a1d7ab" alt="VS Code Installer Help" width="400"/>

</details>

## References

- [VS Code Installer Source (code.iss)](https://github.com/Microsoft/vscode/blob/main/build/win32/code.iss)
- [Inno Setup Command Line Parameters](https://jrsoftware.org/ishelp/index.php?topic=setupcmdline)
- [VS Code Setup Documentation](https://code.visualstudio.com/docs/setup/windows)

---

I hope this helps anyone encountering the same issue. Feel free to add any questions or additional information in the comments.

**Reply by qaisjp** [reply]
llm bot slop

**Comment by lloydjatkinson** [comment]
Are these flags even the correct ones presently? This feels like the sort of thing VS Code team could just change and no one would even know it had changed.

**Reply by emilwojcik93** [reply]
in mine comment somtime ago I put section "[Available VS Code Installer Options](https://github.com/microsoft/winget-cli/discussions/1798#discussioncomment-13355568)" but in general *.iss" is template for Inno Setup generic installer wizard and best practices, recommendations is to use standard flags/params so if something change is most probably it will be related to newer version of Inno Setup template or new fetures for VSCode itself but still it will be announced in changelog and syntax for winget command then should be also updated.

- [VS Code Installer Source (code.iss)](https://github.com/Microsoft/vscode/blob/main/build/win32/code.iss)
- [Inno Setup Command Line Parameters](https://jrsoftware.org/ishelp/index.php?topic=setupcmdline)

**Comment by markembling** [comment]
Anyone else finding this when looking for the answer with a more recent version of WinGet...

The best approach now would be to make use of the `--custom` flag rather than `--override`. That enables specifying of additional flags to the installer without totally overriding those which WinGet applies by default.

So the equivalent now would be:
```
winget install Microsoft.VisualStudioCode --custom '/mergetasks="addcontextmenufiles,addcontextmenufolders"'
```

The `/SILENT`* and `/mergetasks="!runcode"` are part of what WinGet will execute by default (the former based on the installer type declared by the manifest and the latter directly in the manifest as an default additional option). These will remain.

\* I'm actually not sure if it uses `/SILENT` or `/VERYSILENT` but that's sort of the point - we don't need to worry.
